### PR TITLE
Update priority range for PSResourceRepository to 0-100

### LINKED ISSUE #716: Register/Set-PSResourceRepository -Priority range
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PowerShellGet)
- [X] Search the existing issues.

### Steps to reproduce

Currently if you try to register/set a repository to have a lower priority (number higher than 50) than the default PSGallery priority you'll get an error. This is because the parameter only accepts ranges from `0-50`. In the original RFC it was discussed that PSGallery would have a default value of 50 and be neutral priority so you could register repositories higher or lower priority than PSGallery without having to change PSGallery priority.

The second issue is intuitiveness, `-Priority` parameter currently has a type of `[Int32]` which accepts both positive and negative numbers. This is not intuitive to the user as they would need to read the help to understand how the priority system works. However, if you change the type to `[byte]` that only allows values from `0-255` a much more reasonable range. Then PSGallery can be a neutral priority of `100` close enough to halfway of accepted values.

If changing the type is unacceptable my request would be to have the range be `0-100` with newly registered repositories and PSGallery at a neutral priority of `50`.

### Expected behavior

```console
Register-PSResourceRepository -Name Test -Uri c:\temp -Priority 100
```

### Actual behavior

```console
PS C:\> Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
Register-PSResourceRepository: Cannot validate argument on parameter 'Priority'. The 100 argument is greater than the maximum allowed range of 50. Supply an argument that is less than or equal to 50 and then try the command again.
```

### Error details

```console
Exception             :
    Type              : System.Management.Automation.ParameterBindingValidationException
    Message           : Cannot validate argument on parameter 'Priority'. The 100 argument is greater than the maximum
allowed range of 50. Supply an argument that is less than or equal to 50 and then try the command again.
    ParameterName     : Priority
    ParameterType     : int
    TypeSpecified     : int
    ErrorId           : ParameterArgumentValidationError
    Line              : 1
    Offset            : 65
    CommandInvocation :
        MyCommand        : Register-PSResourceRepository
        BoundParameters  :
            Comparer : System.OrdinalIgnoreCaseComparer
            Count    : 2
            Keys     :
                Length : 4

                Length : 3
            Values   :
                Length : 4

                Length : 7
            SyncRoot :
                Comparer : System.OrdinalIgnoreCaseComparer
                Count    : 2
                Keys     :
                    Length : 4

                    Length : 3
                Values   :
                    Length : 4

                    Length : 7
                SyncRoot :
                    Comparer : System.OrdinalIgnoreCaseComparer
                    Count    : 2
                    Keys     :
                        Length : 4

                        Length : 3
                    Values   :
                        Length : 4

                        Length : 7
                    SyncRoot :
                        Comparer : System.OrdinalIgnoreCaseComparer
                        Count    : 2
                        Keys     :
                            Length : 4

                            Length : 3
                        Values   :
                            Length : 4

                            Length : 7
                        SyncRoot :
                            Comparer : System.OrdinalIgnoreCaseComparer
                            Count    : 2
                            Keys     :
                                Length : 4

                                Length : 3
                            Values   :
                                Length : 4

                                Length : 7
                            SyncRoot :
                                Comparer : System.OrdinalIgnoreCaseComparer
                                Count    : 2
                                Keys     :
                                    Length : 4

                                    Length : 3
                                Values   :
                                    Length : 4

                                    Length : 7
                                SyncRoot :
                                    Comparer : System.OrdinalIgnoreCaseComparer
                                    Count    : 2
                                    Keys     : …
                                    Values   : …
                                    SyncRoot : …
        ScriptLineNumber : 1
        OffsetInLine     : 1
        HistoryId        : 45
        Line             : Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
        PositionMessage  : At line:1 char:1
                           + Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
                           + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        InvocationName   : Register-PSResourceRepository
        PipelineLength   : 1
        PipelinePosition : 1
    ErrorRecord       :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Cannot validate argument on parameter 'Priority'. The 100 argument is greater than the maximum
allowed range of 50. Supply an argument that is less than or equal to 50 and then try the command again.
            HResult : -2146233087
        CategoryInfo          : InvalidData: (:) [Register-PSResourceRepository], ParentContainsErrorRecordException
        FullyQualifiedErrorId :
ParameterArgumentValidationError,Microsoft.PowerShell.PowerShellGet.Cmdlets.RegisterPSResourceRepository
        InvocationInfo        :
            MyCommand        : Register-PSResourceRepository
            ScriptLineNumber : 1
            OffsetInLine     : 65
            HistoryId        : 45
            Line             : Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
            PositionMessage  : At line:1 char:65
                               + Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
                               +                                                                 ~~~
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
    TargetSite        :
        Name          : BindParameter
        DeclaringType : System.Management.Automation.ParameterBinderBase, System.Management.Automation,
Version=7.2.5.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Data              : System.Collections.ListDictionaryInternal
    InnerException    :
        Type        : System.Management.Automation.ValidationMetadataException
        ErrorRecord :
            Exception             :
                Type    : System.Management.Automation.ParentContainsErrorRecordException
                Message : The 100 argument is greater than the maximum allowed range of 50. Supply an argument that is
less than or equal to 50 and then try the command again.
                HResult : -2146233087
            CategoryInfo          : MetadataError: (:) [], ParentContainsErrorRecordException
            FullyQualifiedErrorId : ValidateRangeTooBig
        TargetSite  :
            Name          : ValidateRange
            DeclaringType : ValidateRange
            MemberType    : Method
            Module        : System.Management.Automation.dll
        Message     : The 100 argument is greater than the maximum allowed range of 50. Supply an argument that is
less than or equal to 50 and then try the command again.
        Source      : System.Management.Automation
        HResult     : -2146233087
        StackTrace  :
   at System.Management.Automation.ValidateRangeAttribute.ValidateRange(Object element)
   at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter,
CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
    Source            : System.Management.Automation
    HResult           : -2146233087
    StackTrace        :
   at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter,
CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
   at System.Management.Automation.CmdletParameterBinderController.BindParameter(CommandParameterInternal argument,
MergedCompiledCommandParameter parameter, ParameterBindingFlags flags)
   at System.Management.Automation.CmdletParameterBinderController.BindParameter(UInt32 parameterSets,
CommandParameterInternal argument, MergedCompiledCommandParameter parameter, ParameterBindingFlags flags)
   at System.Management.Automation.CmdletParameterBinderController.BindNamedParameter(UInt32 parameterSets,
CommandParameterInternal argument, MergedCompiledCommandParameter parameter)
   at System.Management.Automation.ParameterBinderController.BindNamedParameters(UInt32 parameterSets, Collection`1
arguments)
   at System.Management.Automation.CmdletParameterBinderController.BindCommandLineParametersNoValidation(Collection`1
arguments)
   at System.Management.Automation.CmdletParameterBinderController.BindCommandLineParameters(Collection`1 arguments)
   at System.Management.Automation.CommandProcessor.BindCommandLineParameters()
   at System.Management.Automation.CommandProcessor.Prepare(IDictionary psDefaultParameterValues)
   at System.Management.Automation.CommandProcessorBase.DoPrepare(IDictionary psDefaultParameterValues)
   at System.Management.Automation.Internal.PipelineProcessor.Start(Boolean incomingStream)
   at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
--- End of stack trace from previous location ---
   at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
   at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput,
CommandParameterInternal[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedirection[][]
commandRedirections, FunctionContext funcContext)
   at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : InvalidData: (:) [Register-PSResourceRepository], ParameterBindingValidationException
FullyQualifiedErrorId :
ParameterArgumentValidationError,Microsoft.PowerShell.PowerShellGet.Cmdlets.RegisterPSResourceRepository
InvocationInfo        :
    MyCommand        : Register-PSResourceRepository
    ScriptLineNumber : 1
    OffsetInLine     : 65
    HistoryId        : 45
    Line             : Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
    PositionMessage  : At line:1 char:65
                       + Register-PSResourceRepository -Name test -Uri c:\temp -Priority 100
                       +                                                                 ~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```powershell
Name          Version
----          -------
PowerShellGet 3.0.14

Name                           Value
----                           -----
PSVersion                      7.2.5
PSEdition                      Core
GitCommitId                    7.2.5
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by alerickson**
# PR Summary
Changes the priority range for PSResourceRepository in `Register-PSResourceRepository` and `Set-PSResourceRepository` from 0-50 to 0-100.

The default value remains 50. 

## PR Context
Resolves #716 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: help/Register-PSResourceRepository.md [hunk 1]
```diff
@@ -90,7 +90,7 @@ Accept wildcard characters: False
 
 ### -Priority
 Specifies the priority ranking of the repository.
-Repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40). Has default value of 50.
+Repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40). Has default value of 50.
```
Comment (anamnavi) [inline help/Register-PSResourceRepository.md]:
```suggestion
Repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 80). Has default middle value of 50.
```

File: src/code/RegisterPSResourceRepository.cs [hunk 2]
```diff
@@ -76,12 +76,12 @@ class RegisterPSResourceRepository : PSCmdlet
...
         /// before a lower ranking priority one, when searching for a repository item across multiple registered repositories.
-        /// Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds
+        /// Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds
         /// to a higher priority ranking than a higher numeric value (i.e 40). Has default value of 50.
         /// </summary>
         [Parameter(ParameterSetName = NameParameterSet)]
         [Parameter(ParameterSetName = PSGalleryParameterSet)]
-        [ValidateRange(0, 50)]
+        [ValidateRange(0, 100)]
         public int Priority { get; set; } = defaultPriority;
```
Comment (anamnavi) [inline src/code/RegisterPSResourceRepository.cs]: this is a nitpick but could you do a favor and rename the private fields on line 31 and 32 to `DefaultPriority` and `DefaultTrusted` to match how they are named in Set-PSResourceRepository. Thanks!

File: src/code/RegisterPSResourceRepository.cs [hunk 3]
```diff
@@ -76,12 +76,12 @@ class RegisterPSResourceRepository : PSCmdlet
         /// <summary>
         /// Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched
         /// before a lower ranking priority one, when searching for a repository item across multiple registered repositories.
-        /// Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds
+        /// Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds
         /// to a higher priority ranking than a higher numeric value (i.e 40). Has default value of 50.
```
Comment (anamnavi) [inline src/code/RegisterPSResourceRepository.cs:81]:
```suggestion
        /// to a higher priority ranking than a higher numeric value (i.e 80). Has default value of 50.
```

File: src/code/SetPSResourceRepository.cs [hunk 4]
```diff
@@ -74,12 +74,12 @@ public SwitchParameter Trusted
         /// <summary>
         /// Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched
         /// before a lower ranking priority one, when searching for a repository item across multiple registered repositories.
-        /// Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds
+        /// Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds
         /// to a higher priority ranking than a higher numeric value (i.e 40).
```
Comment (anamnavi) [inline src/code/SetPSResourceRepository.cs:78]:
```suggestion
        /// to a higher priority ranking than a higher numeric value (i.e 80).
```

File: help/en-US/PowerShellGet.dll-Help.xml [hunk 5]
```diff
@@ -2884,7 +2884,7 @@ PS C:\&gt; Get-PSResourceRepository
       <command:parameter required="false" variableLength="true" globbing="false" pipelineInput="False" position="named" aliases="none">
         <maml:name>Priority</maml:name>
         <maml:description>
-          <maml:para>Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40).</maml:para>
+          <maml:para>Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40).</maml:para>
```
Comment (anamnavi) [inline help/en-US/PowerShellGet.dll-Help.xml:2887]:
```suggestion
          <maml:para>Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 80).</maml:para>
```

File: help/en-US/PowerShellGet.dll-Help.xml [hunk 6]
```diff
@@ -2788,7 +2788,7 @@ PS C:\&gt; Get-PSResourceRepository
         <command:parameter required="false" variableLength="true" globbing="false" pipelineInput="False" position="named" aliases="none">
           <maml:name>Priority</maml:name>
           <maml:description>
-            <maml:para>Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40).</maml:para>
+            <maml:para>Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40).</maml:para>
```
Comment (anamnavi) [inline help/en-US/PowerShellGet.dll-Help.xml:2791]:
```suggestion
            <maml:para>Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 80).</maml:para>
```

File: help/en-US/PowerShellGet.dll-Help.xml [hunk 7]
```diff
@@ -1933,7 +1933,7 @@
         <command:parameter required="false" variableLength="true" globbing="false" pipelineInput="False" position="named" aliases="none">
           <maml:name>Priority</maml:name>
           <maml:description>
-            <maml:para>Specifies the priority ranking of the repository. Repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40). Has default value of 50.</maml:para>
+            <maml:para>Specifies the priority ranking of the repository. Repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40). Has default value of 50.</maml:para>
```
Comment (anamnavi) [inline help/en-US/PowerShellGet.dll-Help.xml:1936]:
```suggestion
            <maml:para>Specifies the priority ranking of the repository. Repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 80). Has default value of 50.</maml:para>
```

File: help/Set-PSResourceRepository.md [hunk 8]
```diff
@@ -92,7 +92,7 @@ Accept wildcard characters: False
 ```
 
 ### -Priority
-Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 50, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40).
+Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 40).
```
Comment (anamnavi) [inline help/Set-PSResourceRepository.md]:
```suggestion
Specifies the priority ranking of the repository, such that repositories with higher ranking priority are searched before a lower ranking priority one, when searching for a repository item across multiple registered repositories. Valid priority values range from 0 to 100, such that a lower numeric value (i.e 10) corresponds to a higher priority ranking than a higher numeric value (i.e 80).
```

**Review by anamnavi — APPROVED** [review]
LGTM, just some nitpicks for the comments describing priority lower and higher values to be higher respective to 50.

**Comment by anamnavi** [comment]
Another nitpick, but I think in `RepositorySettings.cs` when we Update ([this line](https://github.com/PowerShell/PowerShellGet/blob/master/src/code/RepositorySettings.cs#L198)) in addition to checking that the new priority is above 0 (meaning it was requested to be set) we should also check that it is under 100 and is valid. So `repoPriority <=0 && repoPriority <= 100)` for the condition. 

We do check that the calling cmdlets ensure it's within 0, 100 but given that these methods are static I think it would help to include this easy check. It would be great if we do this in add too, but I figured in Update where a condition is already happening is easy to include. Thanks!

**Comment by alerickson** [comment]
> LGTM, just some nitpicks for the comments describing priority lower and higher values to be higher respective to 50.

I was confused about those comments... the values don't need to be higher or lower respective to 50, they're just higher or lower respective to each other, PSGallery just happens to have a value of 50.  Was there something in particular you were thinking of in regard to the values being higher/lower than 50?

**Comment by alerickson** [comment]
> Another nitpick, but I think in `RepositorySettings.cs` when we Update ([this line](https://github.com/PowerShell/PowerShellGet/blob/master/src/code/RepositorySettings.cs#L198)) in addition to checking that the new priority is above 0 (meaning it was requested to be set) we should also check that it is under 100 and is valid. So `repoPriority <=0 && repoPriority <= 100)` for the condition.
> 
> We do check that the calling cmdlets ensure it's within 0, 100 but given that these methods are static I think it would help to include this easy check. It would be great if we do this in add too, but I figured in Update where a condition is already happening is easy to include. Thanks!

I have mixed feelings about adding this check in, because it's not necessarily "incorrect" or "wrong" if a repository priority gets updated to a value like say 115, it just gets knocked down to the bottom of the priority list.  However, if a value is less than 0 we assume that it means the user does not want to update their priority, so checking for that does make sense.  Since we're validating that input is between 0 and 100 at the cmdlet level, it theoretically shouldn't be an issue, like you mentioned, but if for some reason a value higher than 100 happened to get added in during an update, it wouldn't break anything or cause any issues.  Just my 2 cents, I'm definitely open to either scenario!

**Comment by alerickson** [comment]
I'm going to merge this PR, but we can continue to discuss and I can open a new PR to address any changes we want to make, since there's still a bunch of small refactoring I'd like to do, like cleaning up error messages, in addition to the larger ones to come.

**Pull request merged**